gui

```diff
@@ -502,49 +458,108 @@ if [[ "$guimode" == yad* ]];then
   trap "rm -f $pipe" EXIT #remove this named pipe on exit
   echo pipe is $pipe
   
-  while true;do #main operating loop for yad window
-    #in background process, send initial app list to yad; first echo line waits until yad is accepting input, then use small delay to prevent items from being missed
-    (echo -e '\f' > $pipe
-    sleep 0.5
-    "${DIRECTORY}/preload" yad '' >> $pipe) &
-    
-    #retrieve a random line from the announcements file for this session
-    motd="$(shuf -n 1 "${DIRECTORY}/data/announcements")"
-    
-    #constantly read from the pipe and send all output to yad
-    #if an item is selected, this script will be run with the row's contents
-    { yad --class Pi-Apps --name "Pi Apps" --center --title='Pi-Apps' \
-      --image="${DIRECTORY}/icons/logo-64.png" --image-on-top --text "$motd" \
-      --list --no-headers --column=:IMG  --column=Name --column=Sysname:HD --column=tip:HD --column=@fore@:HD \
-      --separator='\n' --window-icon="${DIRECTORY}/icons/logo.png" \
-      --tooltip-column=4 \
-      --dclick-action "env pipe=$pipe '${DIRECTORY}/gui'" --select-action "env pipe=$pipe '${DIRECTORY}/gui'" \
-      --button="!${DIRECTORY}/icons/search.png"!'Search':"env pipe=$pipe '${DIRECTORY}/gui' '' '' Search/ Search/" \
-      --button="!${DIRECTORY}/icons/options.png"!'Settings':2 \
-      "$geometry1"; button=$?; kill "$!"; } < <(exec tail -f --retry $pipe)
-    
-    case $button in
-    2) #settings
-      # remove pipe early so that nothing attempts to write to it
-      rm -f "$pipe"
-      "${DIRECTORY}/settings"
-      "${DIRECTORY}/gui" "$@" #run new instance of this script
-      exit 0
-      ;;
+  #in background process, send initial app list to yad; first echo line waits until yad is accepting input, then use small delay to prevent items from being missed
+  (echo -e '\f' > $pipe
+  sleep 0.5
+  "${DIRECTORY}/preload" yad '' >> $pipe) &
+  
+  #retrieve a random line from the announcements file for this session
+  motd="$(shuf -n 1 "${DIRECTORY}/data/announcements")"
+  
+  return_sysname() { #Function that is run in yad gui to echo the selected line to the listening while loop.
+    echo "YAD_XID is $YAD_XID"
+    echo "$3"
+  }
+  
+  refresh_list() { #Refresh the current list of apps in the event of a change
+    if [ ! -z "$pipe" ];then
+      echo -e "\f" > "$pipe"
+      "${DIRECTORY}/preload" yad "$prefix" > "$pipe" 2>/dev/null
+    fi
+  }
+  
+  kill_details() { #close the details window in 1 second. Uses the $detailspid variable. To close it immediately add an argument of 0
+    (
+    if [ "$1" != 0 ];then
+      sleep 1.5
+    fi
```

Comment:
what is the purpose of the sleep? is there a reason why you want to wait to kill the details window when the user clicks on something else?

Reply:
The purpose of this delay is to improve the UX. Previously, when the app details window is open and you select a different app, the first details window will immediately disappear, then about a second later, the new details window appears. If the background behind pi-apps is a dark color, this can cause the overall screen brightness to vary drastically. By aiming to close the old details window after the new one is in place, it makes the experience much more seamless feeling - more like a cohesive UI and less like hacked together dialogs that appear and disappear erratically.

Reply:
> The purpose of this delay is to improve the UX. Previously, when the app details window is open and you select a different app, the first details window will immediately disappear, then about a second later, the new details window appears. If the background behind pi-apps is a dark color but the system GTK theme is light, this can cause the overall screen brightness to vary drastically under normal usage. By aiming to close the old details window after the new one is in place, it makes the experience much more seamless feeling - more like a cohesive UI and less like hacked together dialogs that appear and disappear erratically.

that is because your SD card is slow. on my system with a decent SD card (sandisk A2 512GB, not even the best which is the samsung A2 pro plus), the new details window pops up almost immediately (~ 0.5 seconds) and then the old details window closes afterwards.

I can see how this is a clever way of making sure the background never shows through even if it is weird for me to see the old window close 1.5 seconds later (like if I move it before opening a new one)

Reply:
Fascinating, because I'm booting from almost the fastest NVME drive money can buy, with a USB 3 enclosure. I've always been used to it taking 1-1.5 seconds for just about any type of yad window to appear.
I've long wonder if yad is poorly optimized, or if it waits for something to time out in the background.

Reply:
It could also be Raspbian's Mutter window manager to blame, with its open/close animations.

Reply:
Ultimately it is probably a combination of things.

Mutter is a compositing window manager (uses OpenGL). The pi has a pretty slow gpu and memory.
I use mutter as well (in budgie) but ofc that is with a gpu with a LOT more grunt than the pi.

The pi's ram is very slow (4-5GB/s) and higher latency in comparison to the TX1 (24GB/s).
The pi's very low cpu cache (compared to other chips also with 4 A72 cores) also really hurts its multithreading performance.

yad is also probably poorly optimized. I haven't read its code much (besides to conclude that newer versions are just broken).

Reply:
Also our background shellcheck that runs for 10-15 seconds at 100% of one thread surely doesn't help either 😆